test/python/circuit/test_parameters.py

```diff
@@ -1264,12 +1264,20 @@ def test_parameter_expression_grad(self):
 
         with self.subTest(msg='first order gradient'):
             expr = (x + y) * z
-            self.assertEqual(expr.grad(x), z)
-            self.assertEqual(expr.grad(z), (x + y))
+            self.assertEqual(expr.gradient(x), z)
+            self.assertEqual(expr.gradient(z), (x + y))
+
+        with self.subTest(msg='multiple symbols corresponding to one key'):
```

Comment:
This is still only one symbol, since only one `Parameter` and thus only one `Symbol` is generated. Looking through the rest of the code, I don't think there's a case where there are multiple symbols on one key.
@kdk can you confirm? If so then we can remove the part where you cover the case that the keys are a list.

Reply:
Agree, `_parameter_symbols` should be a one-to-one map from `Parameter` objects to a single `sympy.Symbol`. This is somewhere where our behavior differs from sympy and can be confusing (maybe `_parameter_symbol_map` would be a better name).

Symbols of the same name are equivalent in sympy, but should raise an error for us. (If you had two circuits parameterized each by their own `Parameter('theta')`, and you combined them, you wouldn't necessarily want those `theta`s to be the same.) In the example from above,

```
>>> a = qk.circuit.Parameter('x')
>>> b = qk.circuit.Parameter('y')
>>> c = qk.circuit.Parameter('x')
>>> (a+b)._parameter_symbols
{Parameter(x): x, Parameter(y): y}
>>> (b+c)._parameter_symbols
{Parameter(y): y, Parameter(x): x}  # N.B. The Parameter(x) here != the Parameter(x) above
>>> (a+c)._parameter_symbols
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/parameterexpression.py", line 231, in __add__
    return self._apply_operation(operator.add, other)
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/parameterexpression.py", line 213, in _apply_operation
    self._raise_if_parameter_names_conflict(other._parameter_symbols.keys())
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/parameterexpression.py", line 185, in _raise_if_parameter_names_conflict
    '{}'.format(conflicting_names))
qiskit.circuit.exceptions.CircuitError: "Name conflict applying operation for parameters: {'x'}"
```